### PR TITLE
Add tables pre-fetching to dashboard load

### LINKED ISSUE #44830: [50.x] N+1 queries while loading dashboard `query_metadata`
### Describe the bug

While debugging slow dashboard loads with @LukeAbell yesterday, we found a case of N+1 queries that applies to all dashboard loads. (We found another, worse one #44786 with sandboxing, but this is separate.)

There are 833 queries like this in the logs while loading a midsize dashboard:

```sql
SELECT
  `field`.`base_type`, `field`.`coercion_strategy`, `field`.`database_type`,
  `field`.`description`, `field`.`display_name`, `field`.`effective_type`, `field`.`fingerprint`,
  `field`.`fk_target_field_id`, `field`.`id`, `field`.`name`, `field`.`nfc_path`,
  `field`.`parent_id`, `field`.`position`, `field`.`semantic_type`, `field`.`settings`,
  `field`.`table_id`, `field`.`visibility_type`,
  `dimension`.`human_readable_field_id`, `dimension`.`id`, `dimension`.`name`, `dimension`.`type`,
  `values`.`human_readable_values`, `values`.`values`
FROM `metabase_field` AS `field`
LEFT JOIN `metabase_table` AS `table` ON `field`.`table_id` = `table`.`id`
LEFT JOIN `dimension` AS `dimension` ON
    (`dimension`.`field_id` = `field`.`id`)
    AND `dimension`.`type` IN ('external', 'internal')
LEFT JOIN `metabase_fieldvalues` AS `values` ON
    (`values`.`field_id` = `field`.`id`)
    AND (`values`.`type` = 'full')
WHERE
  (`table`.`db_id` = 7)
  AND (`field`.`id` IN (15051));
```

which is coming from [`lib.metadata.jvm` selecting the pseudotable `:metadata.column`](https://github.com/metabase/metabase/blob/master/src/metabase/lib/metadata/jvm.clj#L135-L137).

I got so far in debugging this as to see that 99% of the calls are coming through the attached stack.
[stacktrace.log](https://github.com/user-attachments/files/16015684/stacktrace.log)

This is a P1 performance issue and the primary reason why `query_metadata` takes 1s instead of 80ms (without sandboxing), but I have to keep debugging the OOMs.

### To Reproduce

Instrument that Toucan call and load pretty much any dashboard.

### Expected behavior

One such call per table, rather than one per field.

(Currently it's also per-query, but I have separate changes queued up to fix that part. Caching these in the `CachedMetadataProvider` is sufficient I think.)

### Logs

_No response_

### Information about your Metabase installation

```JSON
50.x and master both show this.
```

### Severity

P1 perf regression

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by lbrdnk**
Closes https://github.com/metabase/metabase/issues/44830

### Description

This PR is the final part of solution to the https://github.com/metabase/metabase/issues/44830, ie. follow up for https://github.com/metabase/metabase/pull/45050. Analysis of the problem (stacktraces, sql query from linked issue) can be found in [this](https://www.notion.so/metabase/n-1s-while-loading-a-dashboard-7ec125c1d1e547dea132db6eceebdc2b?pvs=4) notion doc.

Briefly, main problem was preprocess being called on every card during dashboard hydration. Thanks to the linked PR metadata providers are now cached, so preprocess metadata is _db queried_ only once for cards with common database.

This PR reduces table db metadata queries. Previously query was executed for every card. This change reduces the amount to one call per database (which dashboard cards _may_ or may not share).

### How to verify

New test was added, counting number of `:metadata/table` toucan calls.

### Checklist

- [X] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: src/metabase/models/card.clj [hunk 1]
```diff
@@ -140,13 +142,39 @@
...
   (let [dataset-cards (filter (comp seq :dataset_query) cards)
         source-card-ids (into #{}
                               (keep (comp source-card-id :dataset_query))
                               dataset-cards)]
+    ;; Prefetching code should not propagate any exception.
+    (try
+      (when lib.metadata.jvm/*metadata-provider-cache*
+        (prefetch-tables-for-cards! dataset-cards))
+      (catch Throwable _
+        (log/errorf "Error prefething cards `%s`." (pr-str (map :id dataset-cards)))))
```
Comment (bshepherdson) [inline src/metabase/models/card.clj:177]: Don't throw away the cause! You can attach the `Throwable` as the first argument to `log/errorf` and we'll log a lot more context.

File: test/metabase/api/dashboard_test.clj [hunk 2]
```diff
@@ -4723,3 +4723,41 @@
...
+    (let [original-select-fn   @#'t2/select
+          uncached-calls-count (atom 0)
+          cached-calls-count   (atom 0)]
+      ;; Get _uncached_ call count of t2/select count for :metadata/table
+      (with-redefs [t2/select (fn [& args]
+                                (when (= :metadata/table (first args))
+                                  (swap! uncached-calls-count inc))
+                                (apply original-select-fn args))]
+        (mt/user-http-request :crowberto :get 200 (format "dashboard/%d" (:id d)))
+        (mt/user-http-request :crowberto :get 200 (format "dashboard/%d/query_metadata" (:id d))))
```
Comment (bshepherdson) [inline test/metabase/api/dashboard_test.clj:4751]: There's already `t2/with-call-count` for exactly this.

**Review by bshepherdson — CHANGES_REQUESTED** [review]

File: src/metabase/models/card.clj (hunk 1 above)
Comment (lbrdnk) [reply src/metabase/models/card.clj]: I believe `(log/errorf t ...)` arity caputures that. Thanks.

File: src/metabase/models/card.clj [hunk 3]
```diff
@@ -140,13 +142,39 @@
...
+  [dataset-cards]
+  (let [db-id->table-ids (-> (group-by :database_id dataset-cards)
+                             (update-vals (partial into #{} (comp (mapcat card->integer-table-ids)
+                                                                  (remove nil?)))))
+        db-id->mp (into {}
+                        (map (juxt identity lib.metadata.jvm/application-database-metadata-provider))
+                        (keys db-id->table-ids))]
+    (doseq [[db-id table-ids] db-id->table-ids
+            :when (seq table-ids)]
+      (lib.metadata.protocols/metadatas (db-id->mp db-id) :metadata/table table-ids))))
```
Comment (bshepherdson) [inline src/metabase/models/card.clj:163]: Defining `db-id->mp` as a separate map is redundant here - each one is only used in one iteration of the `doseq`, which is already grouped by `db-id`.

```suggestion
  (let [db-id->table-ids (-> (group-by :database_id dataset-cards)
                             (update-vals (partial into #{} (comp (mapcat card->integer-table-ids)
                                                                  (remove nil?)))))]
    (doseq [[db-id table-ids] db-id->table-ids
            :let  [mp (lib.metadata.jvm/application-database-metadata-provider db-id)]
            :when (seq table-ids)]
      (lib.metadata.protocols/metadatas mp :metadata/table table-ids))))
```

File: test/metabase/api/dashboard_test.clj (hunk 2 above)
Comment (bshepherdson) [reply test/metabase/api/dashboard_test.clj:4764]: That's fair.

**Review by bshepherdson — APPROVED** [review]

File: src/metabase/models/card.clj (hunk 3 above)
Comment (lbrdnk) [reply src/metabase/models/card.clj]: Thanks for the suggestion. Applied!

**Pull request merged**